apps/emqx_gateway/src/emqx_gateway_http.erl

```diff
@@ -0,0 +1,208 @@
+                  || Node <- ekka_mnesia:running_nodes()]).
+
+lookup_client(Node, GwName, {clientid, ClientId}, {M,F}) when Node =:= node() ->
+    ChanTab = emqx_gateway_cm:tabname(chan, GwName),
+    InfoTab = emqx_gateway_cm:tabname(info, GwName),
+
+    lists:append(lists:map(
+      fun(Key) ->
+        lists:map(fun M:F/1, ets:lookup(InfoTab, Key))
+      end, ets:lookup(ChanTab, ClientId)));
+
+lookup_client(Node, GwName, {clientid, ClientId}, FormatFun) ->
+    rpc_call(Node, lookup_client,
+             [Node, GwName, {clientid, ClientId}, FormatFun]).
+
+-spec kickout_client(gateway_name(), emqx_type:clientid())
+    -> {error, any()}
+     | ok.
+kickout_client(GwName, ClientId) ->
+    Results = [kickout_client(Node, GwName, ClientId)
+               || Node <- ekka_mnesia:running_nodes()],
+    case lists:any(fun(Item) -> Item =:= ok end, Results) of
+        true  -> ok;
+        false -> lists:last(Results)
+    end.
+
+kickout_client(Node, GwName, ClientId) when Node =:= node() ->
+    emqx_gateway_cm:kick_session(GwName, ClientId);
+
+kickout_client(Node, GwName, ClientId) ->
+    rpc_call(Node, kickout_client, [Node, GwName, ClientId]).
+
+-spec list_client_subscriptions(gateway_name(), emqx_type:clientid())
+    -> {error, any()}
+     | {ok, list()}.
+list_client_subscriptions(GwName, ClientId) ->
+    %% Get the subscriptions from session-info
+    with_channel(GwName, ClientId,
+        fun(Pid) ->
+            Subs = emqx_gateway_conn:call(
+                     Pid, 
+                     subscriptions, ?DEFAULT_CALL_TIMEOUT),
+            {ok, lists:map(fun({Topic, SubOpts}) ->
+                     SubOpts#{topic => Topic}
+                 end, Subs)}
+        end).
+
+-spec client_subscribe(gateway_name(), emqx_type:clientid(),
+                       emqx_type:topic(), emqx_type:subopts())
+    -> {error, any()}
+     | ok.
+client_subscribe(GwName, ClientId, Topic, SubOpts) ->
+    with_channel(GwName, ClientId,
+        fun(Pid) ->
+            emqx_gateway_conn:call(
+              Pid, {subscribe, Topic, SubOpts},
+              ?DEFAULT_CALL_TIMEOUT
+             )
+        end).
+
+-spec client_unsubscribe(gateway_name(),
+                         emqx_type:clientid(), emqx_type:topic())
+    -> {error, any()}
+     | ok.
+client_unsubscribe(GwName, ClientId, Topic) ->
+    with_channel(GwName, ClientId,
+        fun(Pid) ->
+            emqx_gateway_conn:call(
+              Pid, {unsubscribe, Topic}, ?DEFAULT_CALL_TIMEOUT)
+        end).
+
+with_channel(GwName, ClientId, Fun) ->
+    case emqx_gateway_cm:with_channel(GwName, ClientId, Fun) of
+        undefined -> {error, not_found};
+        Res -> Res
+    end.
+
+%%--------------------------------------------------------------------
+%% Utils
+%%--------------------------------------------------------------------
+
+-spec return_http_error(integer(), binary()) -> {integer(), binary()}.
+return_http_error(Code, Msg) ->
```

Comment:
already defined response {ErrorHttpCode :: integer() , Code :: atom(), Message :: binary()} by minirest tag 1.2.1

Reply:
Great. I will replace it in the future